powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/totp/TOTP.java

```diff
@@ -0,0 +1,299 @@
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Generates a TOTP value for the given set of parameters.
+     *
+     * @param key          the shared secret
+     * @param timeStep     number of time step
+     * @param returnDigits number of digits to return
+     * @param algorithm    the algorithm to use
+     * @return a numeric String in base 10 that includes truncation digits
+     * @throws CryptoProviderException in case of any crypto error
+     */
+    private static String generateTOTP(final byte[] key, final long timeStep, final int returnDigits, final String algorithm) throws CryptoProviderException {
+        logger.debug("Generating TOTP for timeStep={}, algorithm={}", timeStep, algorithm);
+
+        if (key == null) {
+            throw new CryptoProviderException("Key is mandatory");
+        }
+
+        if (algorithm == null) {
+            throw new CryptoProviderException("Algorithm is mandatory");
+        }
+
+        if (returnDigits <= 0) {
+            throw new CryptoProviderException("ReturnDigits must be positive number");
+        }
+
+        // Using the counter
+        // First 8 bytes are for the movingFactor
+        // Compliant with base RFC4226 (HOTP)
+        final String timeWithCounterPrefix = padWithZeros(Long.toHexString(timeStep), 16);
+
+        // Get the HEX in a Byte[]
+        final byte[] msg = HexFormat.of().parseHex(timeWithCounterPrefix);
+
+        final byte[] hash = computeHash(algorithm, key, msg);
+
+        // put selected bytes into result int
+        final int offset = hash[hash.length - 1] & 0xf;
+
+        final int binary = ((hash[offset] & 0x7f) << 24) |
+                ((hash[offset + 1] & 0xff) << 16) |
+                ((hash[offset + 2] & 0xff) << 8) |
+                (hash[offset + 3] & 0xff);
+
+        final int otp = binary % DIGITS_POWER[returnDigits];
+
+        final String result = Integer.toString(otp);
+        return padWithZeros(result, returnDigits);
+    }
+
+    private static long countTimeSteps(final LocalDateTime localDateTime) throws CryptoProviderException {
+        if (localDateTime == null) {
+            throw new CryptoProviderException("LocalDateTime is mandatory");
+        }
+
+        return localDateTime.toEpochSecond(ZoneOffset.UTC) / TIME_STEP_X;
+    }
+
+    private static String padWithZeros(final String source, final int length) {
+        return Strings.padStart(source, length, '0');
+    }
+
+    /**
+     * Computes a Hashed Message Authentication Code with the give hash algorithm as a parameter.
+     *
+     * @param algorithm the algorithm
+     * @param keyBytes  the bytes to use for the HMAC key
+     * @param text      the message or text to be authenticated
+     * @throws CryptoProviderException in case of any crypto error
+     */
+    private static byte[] computeHash(final String algorithm, final byte[] keyBytes, final byte[] text) throws CryptoProviderException {
```

Comment:
The variable name `text` is a bit confusing, because it is of `byte[]` type. I would use something like `input` or `data`, not `text`.